flow/flow.cpp

```diff
@@ -21,9 +21,22 @@
 #include "flow/flow.h"
 #include "flow/DeterministicRandom.h"
 #include "flow/UnitTest.h"
+#include "flow/rte_memcpy.h"
+#include "flow/folly_memcpy.h"
 #include <stdarg.h>
 #include <cinttypes>
 
+// For benchmarking; need a version of rte_memcpy that doesn't live in the same compilation unit as the test.
+void * rte_memcpy_noinline(void *__restrict __dest, const void *__restrict __src, size_t __n) {
+	return rte_memcpy(__dest, __src, __n);
+}
+
+// This compilation unit will be linked in to the main binary, so this should override glibc memcpy
```

Comment:
Does this compile on ARM?
Also, IMHO, we should have a memcpy pointer and fix the pointer at runtime.
So we can easily add a new memcpy implementation when the next AVX comes.

Reply:
No; it doesn't compile on ARM (yet).  This should have been marked as a draft.

I'm looking into using the `ifunc` mechanism to do the runtime selection of memcpy (and hopefully other functions:

https://gcc.gnu.org/onlinedocs/gcc-4.7.2/gcc/Function-Attributes.html

It uses some ELF magic to implement the function pointer idea you're describing (presumably with slightly lower overhead).

Reply:
Ah, sorry, got confused with folly_memcpy.h and rte_memcpy.h (which was hidden in my browser).  I think dispatching in rte_memcpy.h works.  Although we still need to handle other platforms, I suppose.

Reply:
... and my browser just showed your reply after I posted the second comment!  Sorry!

Reply:
Do we have an easy way to check if it compiles on ARM?  This change should break the ARM build (and I'd like to update it to make sure it doesn't).

Reply:
I don't know what's easy 😉.  The easiest way is probably just build it on an ARM machine...
I believe it's fairly easy to do cross compilation with clang, but I've never done it.
@sfc-gh-mpilman might have a better idea on clang cross-compilation, maybe?

Reply:
For now this memcpy could also just be guarded by cmake. Basically we could use default `memcpy` on ARM and then figure out what to do.

Reply:
I added a guard based on the availability of __AVX__.  I can test that without cross-compiling.